Change remaining Coil:Heating:Gas to Fuel plus cleanups on new coil sizing reports

mjwitte opened the pull request:
Pull request overview
---------------------
Part 2 of new coil sizing reports, completing #6454.

 - [x] Fix plant index errors
 - [x] Add design document
 - [x] Finish documentation
 - [x] Add unit tests
 - [x] Refine units in column headings - use standard units enums or equivalent?
 - [x] Add timestamp tag to date/time columns

Also . . .

- [x] Fixed lingering occurrences of Coil:Heating:Gas --> Fuel, which actually were discovered by one of these new unit tests (Coil:Heating:Fuel wasn't being recognized as a heating coil)
- [x] Noted ip-units conversion warnings for table outputs and fixed units in the Surface Convection Parameters eio output header.
- [x] Partially addresses #6481 

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

mjwitte committed:
Coil sizing report - fix plant index errors

mjwitte committed:
Coil sizing report - design document

mjwitte committed:
Coil sizing report - I/O Ref changes add report to an example file

mjwitte committed:
Coil sizing report - Initial unit test

Myoldmopar commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1044,7 +1044,48 @@ ReportCoilSelection::setCoilAirFlow(
+)
+{
+	int index = getIndexForOrCreateDataObjFromCoilName( coilName,coilType );
+	auto & c( coilSelectionDataObjs[ index ] );
```
👍

Myoldmopar commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1044,7 +1044,48 @@ ReportCoilSelection::setCoilAirFlow(
+	if ( c->waterLoopNum > 0 && c->pltSizNum > 0 ) {
+		c->plantLoopName = DataPlant::PlantLoop( c->waterLoopNum ).Name;
+		if ( DataSizing::PlantSizData( c->pltSizNum  ).LoopType != DataSizing::SteamLoop ) {
+			c->rhoFluid = FluidProperties::GetDensityGlycol( DataPlant::PlantLoop( c->waterLoopNum ).FluidName, DataGlobals::InitConvTemp, DataPlant::PlantLoop( c->waterLoopNum ).FluidIndex, "ReportCoilSelection::setCoilWaterFlow" );
```
Thanks for being explicit on the scope of these functions and not using using statements!

Myoldmopar commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1044,7 +1044,48 @@ ReportCoilSelection::setCoilAirFlow(
+	if ( isAutoSized ) {
+		c->coilWaterFlowAutoMsg = "Yes" ;
+	} else {
+		c->coilWaterFlowAutoMsg = "No" ;
```
So easy to read all this with the `c` reference variable.

Myoldmopar commented on src/EnergyPlus/ReportCoilSelection.hh:
```diff
@@ -374,17 +385,29 @@ public: // methods
 		Real64 const totalHeatingCap, // {W} coil Heating capacity
 		bool const isAutoSize, // true if value was autosized
-		int const dataPltSizNum, // plant sizing structure index
+		int const DataPltSizNum, // plant sizing structure index
```
Was there a reason for capitalizing this?

mjwitte replied:
Hmm, don't remember changing  that.

Myoldmopar commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,96 @@
+	Real64 airVdot ( 0.052 ); // air flow rate in m3/s
+	bool isAutoSized ( false ); // true if air flow was autosized
+
+	coilSelectionReportObj->setCoilAirFlow( coilName, coilType, airVdot, isAutoSized );
```
So concise.  Are you adding more unit tests or is this the extent of your testing?

Myoldmopar replied:
From the [coverage results](http://energyplus.s3-website-us-east-1.amazonaws.com/coverage/2018-02-27-EnergyPlus-4b97a62bb0-x86_64-gcc-4.8UTCD/EnergyPlus/ReportCoilSelection.cc.gcov.html) it looks like the code that isn't covered by this test is mostly the erroneous code paths and the steam loop calculations.  However there is a section in the `setCoilWaterFlowNodeNums` function where `c->waterLoopNum > 0 && c->pltSizNum > 0` and that block doesn't get hit.  What do you think about adding some coverage there?

mjwitte replied:
This is only half-baked. Will add more.  The functions themselves are all from @EnergyArchmage so I can't take credit for how they're written. I just made copies of them to make two flavors for some.

Myoldmopar reviewed (commented):
This looks really good.  I had a question about the unit tests.  Let me know what you think there and we'll get it approved.

mjwitte commented on src/EnergyPlus/ReportCoilSelection.hh:
```diff
@@ -470,7 +493,7 @@ private: // methods
 
 
-private: // data
+public: // data
```
@Myoldmopar @EnergyArchmage I couldn't get the unit tests to work without making this public. Is there a better way?

Myoldmopar replied:
https://stackoverflow.com/questions/2396370/how-to-make-google-test-classes-friends-with-my-classes

I think make it public.  Or better yet make your `class` a `struct` and just make things private that are definitely not needed anywhere else even in unit tests.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into CoilSizingFollowup

# Conflicts:
#	src/EnergyPlus/WaterCoils.cc

mjwitte committed:
Coil sizing report - fix build warnings and merge problems

mjwitte committed:
Coil sizing report - modes to setCoilWaterFlow* and more unit tests

mjwitte commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,365 @@
+	coilSelectionReportObj->setCoilAirFlow( coil1Name, coil1Type, airVdot, isAutoSized );
+	auto & c1 ( coilSelectionReportObj->coilSelectionDataObjs[ 0 ] );
+	EXPECT_EQ( coil1Name, c1->coilName_ );
+	EXPECT_EQ( coil1Type, c1->coilObjName );
```
@Myoldmopar This works here and I can query c1 and see that's filled where it should be. But later on after coil 2 is created, c1 becomes undefined. What needs to be different here when declaring c1 to make it stick later?

mjwitte commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,365 @@
+	// First with no plant sizing objects defined
+	isAutoSized = false; // true if autosized
+	coilSelectionReportObj->setCoilWaterFlowNodeNums( coil1Name, coil1Type, waterVdot,  isAutoSized, chWInletNodeNum, chWOutletNodeNum, loopNum );
+	EXPECT_EQ( -999, c1->pltSizNum );
```
c1 is still ok here.

mjwitte commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,365 @@
+	std::string coil2Type ( "Coil:Cooling:Water" ); // idf input object class name of coil
+	int pltSizNum = -999;
+	coilSelectionReportObj->setCoilWaterFlowPltSizNum( coil2Name, coil2Type, waterVdot,  isAutoSized, pltSizNum, loopNum );
+	auto & c2 ( coilSelectionReportObj->coilSelectionDataObjs[ 1 ] );
```
But here after a second coilSelectionDataObjs is created, c1 is undefined.

mjwitte commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,365 @@
+	DataSizing::PlantSizData( 1 ).PlantLoopName = "Chilled Water Loop";
+	isAutoSized = true; // true if autosized
+	coilSelectionReportObj->setCoilWaterFlowNodeNums( coil1Name, coil1Type, waterVdot,  isAutoSized, chWInletNodeNum, chWOutletNodeNum, loopNum );
+	auto & c1b ( coilSelectionReportObj->coilSelectionDataObjs[ 0 ] );
```
So, here for now, creating a new c1b to point to the first coil.

EnergyArchmage committed:
coil Summary report corrections

change zone weighting to supply-air-flow-rate from zone volume, traps to
not overwrite values set directly

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into CoilSizingFollowup

mjwitte committed:
Coil sizing report - more unit test coverage

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into CoilSizingFollowup

Myoldmopar commented:
This looks really good.  I am happy to see the plant index errors gone.  Looking at the diffs, they seem to be only in the table, and spot checking those I see differences in coil sizes that are relatively small, but numerically significant for throwing diffs.  I have no issue with this.  Were you planning on doing anything else on here?  If not I'll merge.

mjwitte commented:
@Myoldmopar Just about finished with expanded unit tests and still want to clean up some units in the table headers. Should be up within the hour.

Myoldmopar commented:
:+1:

mjwitte committed:
Fix lingering Coil:Heating:Gas --> Coil:Heating:Fuel

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into CoilSizingFollowup

mjwitte committed:
Coil sizing report - more unit test coverage again

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into CoilSizingFollowup

mjwitte committed:
Coil sizing report - table heading cleanup

mjwitte committed:
Fix units in eio Surface Convection Parameters heading

mjwitte commented:
@Myoldmopar Note change in labels, title and new items in overview.

Myoldmopar commented:
Thanks for covering this so well @mjwitte.  I'm going to avoid merging for a little while and let CI have a few hours to check on diffs in this and a couple others that are basically ready to go.  If this is clean it'll go in.

mjwitte commented:
@Myoldmopar Arrrgh. Tablediff is [failing now](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-e7ccfff88223a238bd5d66cc022f021f5c48a4e7-x86_64-MacOS-10.9-clang.html).
(diff, which) = hdict[h][i]
KeyError: u'Date/Time at Sensible Ideal Loads Peak'

mjwitte commented on src/EnergyPlus/OutputReportPredefined.cc:
```diff
@@ -1972,25 +1972,25 @@ namespace OutputReportPredefined {
 		//results from regular zone and system sizing calcs, "At Ideal Loads Peak"
 		pdch2CoilDDnameSensIdealPeak = newPreDefColumn( pdst2CoilSummaryCoilSelection, "Design Day Name at Sensible Ideal Loads Peak" );
-		pdch2CoilDateTimeSensIdealPeak = newPreDefColumn( pdst2CoilSummaryCoilSelection, "Date/Time at Sensible Ideal Loads Peak" );
+		pdch2CoilDateTimeSensIdealPeak = newPreDefColumn( pdst2CoilSummaryCoilSelection, "Date/Time at Sensible Ideal Loads Peak {TIMESTAMP}" );
```
@JasonGlazer So, adding {TIMESTAMP} here is causing TableDiff to [crash](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-e7ccfff88223a238bd5d66cc022f021f5c48a4e7-x86_64-MacOS-10.9-clang.html). Can you see anything wrong here?

JasonGlazer replied:
Nothing jumps out as to why this would cause a problem.

mjwitte committed:
Coil sizing report - revert table heading {TIMESTAMP} tags

mjwitte commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -0,0 +1,427 @@
+
+	Real64 zoneHeatingLatentLoad = 4321.0;
+	coilSelectionReportObj->setZoneLatentLoadHeatingIdealPeak( zoneNum, zoneHeatingLatentLoad );
+	// Expect zero because doZoneEqSetup isn't currently executed
```
@EnergyArchmage I noticed that all calls to coilSelectionReportObj->doZoneEqSetup are currently commented out. Is that your intent at this point?

Myoldmopar commented:
Hmm....looks like it's still failing with TableDiff?

mjwitte commented:
@JasonGlazer What the . . . can TableDiff no handle *any* change in a column heading?

mjwitte committed:
Coil sizing report - revert table headings to KGWATER/KGDRYAIR

mjwitte commented:
One more time . . . hope TableDiff is happy now.

Myoldmopar commented:
Back to tablediffs, not tablediff failures!  Ready to go once CI continues a bit more?

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CoilSizingFollowup

Myoldmopar commented:
Pulled in the latest develop; letting CI run overnight and we'll merge in the morning.

Myoldmopar commented:
Clang was not pleased with this last commit, but it could be a fluke.  I just deleted those results and we'll see if Marvin is happier on another try.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CoilSizingFollowup

Myoldmopar commented:
I merged in a couple more PRs so I went ahead and pulled develop in and pushed it back up.  With the drastically reduced number of branches CI is watching, it should get done with a full set of diffs relatively soon this morning.

Myoldmopar commented:
This looks to be good to go but I'm going to hold off from merging so the diffs don't get intermingled over on #6498.  I'll merge this once the base Linux CI over there is done.  Then we'll be nearly done...
